### ISSUE TITLE
Implement New Invite System

### ISSUE THREAD

**Issue by defnax** [RetroShare/RetroShare#638]
#**New Invite System**

Allow one way Add Friend via Key, on other side Friend gets Friend Reqeust (dont need both sides to add key)Not sure when this will be implemented? use DHT?
If we develope a RetroShare for all, 12 to 60 years old people must be able to use it, it must be easy to reach more people with RetroShare.

**Introduction**
This page is the Overview for a Proposed Invitation System for RetroShare.
It is supposed to list the desired requirements for the System, and how closely the proposals meet these requirements.

**Use Cases**
1 - Two people find retroshare, and want to use it.
2 - An existing retroshare user wants to invite a friend to use retroshare.
3 - two retroshare users (unconnected via gpg graph) want to be friends.
4 - two retroshare users (connected via gpg graph) want to be friends.
5 - a retroshare user wants to introduce two friends to each other.
6 - a retroshare user wants an introduction to a friend of a friend.

Requirements

**Mandatory**
1) Authentication Reliable

**Essential**
1) Ability to attach a personal message.
2) Simple to Use.
3) Works for Firewalled Peers.
4) Works for Offline Peers
5) Works with Unknown Peers.
6) Works with Friends of Friends.
7) Indication if Invitation impossible

**Desirable**
1) Unique Code to indicate Invitation.
2) Small UniqueCode Size.
3) Single Click.
4) Invitation Times Out.
5) Invitation Completed Immediately (No Delay between Invitation Received and Signing Key)
6) Allows Introductions. 

 **Proposals Overview** 
![rs](https://cloud.githubusercontent.com/assets/9952056/21830717/f1f5ccec-d79e-11e6-830d-c742f1803c47.png)

**## Invitation system: DHT IP STORE**

For me the invitation should be an easy way to make friends, while keeping the key signature paradigm.

**Two typical cases:**

1 - A adds a new friend B that does not understand much about gpg, signatures, etc. In such a case the "invite friend" button should propose to generate a small string/passwd/whatever that A sends to B

2 - A wants to invite a peer from his network tab, and let him know

I think it's possible to handle both cases this way:

**A wants to make friend with B**

-> A generates an invitation code.
-> B gets the code, add it to RS. RS pops a window at B, asking to sign A's key
-> A gets informed that B did. RS pops a window at A, asking to sign B's key

Depending on the configuration (1 or 2, as listed above) some of these steps might be skipped

**Proposed Algorithm**
One option for implementing this is:

1- The invitation code from A is a hash. A posts to the DHT this hash with his IP
2- B gets the code and looks into the DHT for A's IP. When B has it, it replaces the DHT entry with his own IP
3- A keeps looking into the DHT for a new IP, and find B's
4- at the same time, both peers send connexion requests to each other in a clear packet with public GPG key
5- each RS tells the other "this peers with this GPG key/name/ID" is requesting to be friend with you, do you accept ?
6- A and B sign each other's keys.

In case 2, if A wants to invite B which key is already in his network tab, the invitation starts at step 4 with asymmetric invitation: A sends to B, and B replies (if he accepts the invitation) with an invitation packet containing his public key.

Invitations should have a time limit, for security, and might be used only once, accept friends automatically, etc depending on what A chooses, as everything is handled locally.

A passwd might be used to encrypt the IPs in the DHT, so that the invitation is optionaly made more secure.

Options should be made available so as to only propose invitations responses to already sent invitations (case 1 for A only) or popup a window for all invitations (case 2 for A), etc.

**Comments On This Proposal**
**Cyril:** _This system needs peers to be able to exchange clear packets to send public keys. This needs to be cleanly implemented to avoid any possible security flaw. Still, this method has some definite advantages:

the invitation code needs to be exchanged on way only. RS handles all the relevant traffic such as public GPG key exchange, peer ids, IPs, etc.
it suppressed the need for explicit GPG key exchange, which confuses most non-geek users._ 

**DrBob:** _This idea looks great on paper, but has several flaws when examined in detail:

The biggest issue is that there will be a big time gap between the initial "Send Code", and actually receiving the peers GPG key (allowing the "Accept Friend" to take place). This will be very confusing for users. This will happen because both peers are required to be simultaneously online before the Key Exchange can occur, and as we all know, sometimes it can be days, weeks or months before two people happen to be simultaneously online. 

The Bittorrent DHT is not designed to store generic Hashes. We would have to add an RS extension, but would not be functional until we have sufficient RS DHT peers (~100-1000). This problem is solvable, by directly searching for the peer matching the hash (how RS finds DHT peers now) 

An open "invitation port", would allow random peers to trivially identify every RS node. All they would have to do is send a random GPG Key to the port... and you would get the Peers Key returned. This problem is solvable, by composing the "INVITE CODE" of two parts: "DHT HASH" + "UNIQUE AUTH CODE", and only communicating with valid Auth Codes. 

Due to the firewalled nature of a large portion of the internet, this invitation system will fail for any Peer B that doesn't have an open external port. As an example, we can see that this is typically 2/3 of all DHT nodes are firewalled. Imagine that RS is twice as good... this means 1/3 of all invitations will 'just not work' - which is pretty bad. (For comparison, if both peers are firewalled, then an RS connection is very difficult ... (1/3)^2 = 11.1% of the time). This problem is very difficult to get around. 

Authentication is difficult if the peers are not part of the existing GPG network. All you are relying on is the "INVITE CODE", which seems pretty weak to me! 

 This invitation system requires a whole new "invitation port" subsystem to be built. Quite a lot of work._ 

7 years ago it was idea but never implemented
Source:
http://retroshare.sourceforge.net/wiki/index.php/Invitation_System_Proposals
http://retroshare.sourceforge.net/wiki/index.php/Invite_Proposal_1,_DHT_IP_STORE
http://retroshare.sourceforge.net/wiki/index.php/INVITE_PROP2_SERVICE

**Comment by defnax** [comment]
Look here Tox IM routs Friend Requests:
![projecttox-free-as-in-freedom-skype-replacement-45-638](https://cloud.githubusercontent.com/assets/9952056/21831059/67fb6454-d7a1-11e6-8665-a66282664959.jpg)

**Comment by defnax** [comment]
Infos how in tox fr works:
https://toktok.ltd/spec#friend-requests

**Comment by MikeMike4R** [comment]
I agree ease of use is integral.
Could we chat by email.
mm4r@protonmail.com

**Comment by szh7379** [comment]
Being expecting it, It will be a milestone.

**Comment by gongzisun** [comment]
Excpecting too. Thanks to all the programmer.

**Comment by ASmith-** [comment]
This coding appeared to already exist for sometime in the 'Mail' tool and retroshare source code. Then recently, it suddenly was removed. No discussion, not even a hint that would be eliminated. Up to that point I and many others had used that feature hundreds of times.

Previous coding tied to the Mail Tool.
A user would add your key from a chat lobby link or from the key exchange forums and the 'then' coding would discover two things. That public key of the person matched your public key but the remote peer's public key isn't on your keyring. (in the background a temporary copy of the distant peers public key was copied) A personalized Mail would then be generated telling the user this person wants to be added as a friend. Button 'Add this Person' .... CLICK. That temporary copy would then be added to your public keyring. DONE.

All apparently gutted, perhaps by someone changing the coding for Android or a new GXS Mail system, who knows, and those that know should be talking about this before it was gutted.

'Implement New Invite System' really sounds much as already previously existed and certainly wouldn't be all that problematic for the party gutting that recently and returning it.

**Comment by G10h4ck** [comment]
short invite system has been implemented in master and will be available for next release

**Comment by defnax** [comment]
> short invite system has been implemented in master and will be available for next release

i speak about a invite system when i accept the key , then send to other side the friend request to accept to get friends.
I think that is not done only key is short now right?

**Comment by G10h4ck** [comment]
With my initial proposal it was possible since one could add a friend just by knowing his own SSL-id that is available when connection is attempted. I have just checked the current PR that @csoler built on top of my initial work and it is not possible anymore since now to add a friend also PGP-fingerprint is needed which is not available at connection attempt (while PGP-id is available). So either so change is made or the comfortable "Accept" button is not possible to implement anymore -_-

**Comment by G10h4ck** [comment]
IMHO the possibility to add someone as friend on attempt is a really important usability feature, even in one of the most common social interaction people ask your phone number, and then give you a miss call so you have their number instead of blabbering it to you...

So I have double checked the code in PR https://github.com/RetroShare/RetroShare/pull/1571 , and it is still possible to implement an accept button after a failed connection attempt notification at user interface level.

Basically when someone adds you through an invite you provided, she will attempt connections towards you, if don't have added her yet your RS will deny the connection and an event is posted see `AuthSSLimpl::VerifyX509Callback` the event contains the data such as SSL-id and PGP-id we learnt from the connection.

With that information the UI as an example of a possible implementation, could display an "Accept" button together with the notification, when "Accept" button is pressed `RsPeers::addSslOnlyFriend` can be called with the info contained in the event,

**Comment by csoler** [comment]
PR1571 does not remove the possibility to add a friend on connection attempt. That possibility is not (and has never been) offered when you actually do not have the PGP key of the peer trying to contact you.

The only problem is that when you read the PGP id from the handshake at a connection attempt, you may not have this key in your keyring. Before (and still after PR 1571) the possibility to add a friend attempting to connect, for which we don't know the PGP key, is not offered.

We could obviously add this feature, but PR 1571 doesn't remove anything that was already here.

**Comment by csoler** [comment]
> > short invite system has been implemented in master and will be available for next release
> 
> i speak about a invite system when i accept the key , then send to other side the friend request to accept to get friends.
> I think that is not done only key is short now right?

PR 1571 doesn't do that. It just shortens the certificates without reducing security.
It is however possible to use the same trick we use in PR 1571, in order to allow users to add friends on connection request even if the PGP public key is not available. The only step toward this is to use the full PGP fingerprint in the handshake instead of the 8-bytes ID.
